Fix invalid group problems & error reporting on startup

bartico6 opened the pull request:
Server will no longer start up without the two essential groups (default and guest).
![](https://i.quake.tf/1626205229-29014585.png)

Additionally, players whose groups are removed in the database will no longer be stuck on connecting phase, and are instead disconnected with a descriptive message.
![](https://i.quake.tf/1626205140-48536470.png)

The server now also correctly reports startup errors in the console. The previous call to 'Error' instead of 'ConsoleError' caused the error to appear in the log, but not in the console, leading to a confusing situation for both users and developers where the server would randomly terminate mid-startup.

Fixes #2396

bartico6 committed:
Missing group safeguards.

- Server will no longer start up when the guest or default groups cannot
  be located.
- Players joining with unknown groups assigned to them will be
  disconnected with an error

bartico6 committed:
Fix the server not reporting startup errors correctly.

bartico6 committed:
Update CHANGELOG.md

bartico6 committed:
/login kick -> error, add XML doc to Utils method.

bartico6 committed:
Minor update.

- AssertGroupValid now both sends the message and kicks the player
  depending on input parameter.
- /login and DataHandler code is now an identical assert check.

hakusaro commented on TShockAPI/Utils.cs:
```diff
@@ -212,6 +212,27 @@ public int GetActivePlayerCount()
+		/// <param name="group"></param>
+		/// <returns></returns>
+		public bool AssertGroupValid(TSPlayer player, Group group, bool kick)
+        {
```
Can you indent this with tabs?

hakusaro commented on CHANGELOG.md:
```diff
@@ -29,6 +29,7 @@ This is the rolling changelog for TShock for Terraria. Use past tense when addin
 * Correct rejection message in LandGolfBallInCupHandler to output the proper expected player id. (@drunderscore)
 * Clarified the error mesage that the console is presented if a rate-limit is reached over REST to indicate that "tokens" actually refers to rate-limit tokens, and not auth tokens, and added a hint as to what config setting determines this. (@hakusaro, @patsore)
 * Fixed an issue where, when the console was redirected, input was disabled and commands didn't work, in TSAPI. You can now pass `-disable-commands` to disable the input thread, but by default, it will be enabled. Fixes [#1450](https://github.com/Pryaxis/TShock/issues/1450). (@DeathCradle, @QuiCM)
+* Fixed missing group problems & error reporting on startup. (@bartico6)
```
Can you be as specific and descriptive as possible with how you did this?

hakusaro commented on TShockAPI/Utils.cs:
```diff
@@ -212,6 +212,27 @@ public int GetActivePlayerCount()
+		/// <param name="player"></param>
+		/// <param name="group"></param>
+		/// <returns></returns>
+		public bool AssertGroupValid(TSPlayer player, Group group, bool kick)
```
Can you move this into the group manager place/thing? We're trying not to add anything else into utils, because it's kinda just the trash dumping ground. I don't want you to just chuck something in, because then it has to live with the other `Dump` methods, which really make this look kinda...redheaded.

hakusaro commented on TShockAPI/TShock.cs:
```diff
@@ -322,6 +322,7 @@ public override void Initialize()
 				Regions = new RegionManager(DB);
 				UserAccounts = new UserAccountManager(DB);
 				Groups = new GroupManager(DB);
+				Groups.EnsureCoreGroupsPresent();
```
Couldn't this happen on the constructor for the GroupManager?

hakusaro commented on TShockAPI/DB/GroupManager.cs:
```diff
@@ -202,6 +202,21 @@ public GroupManager(IDbConnection db)
+            {
+				TShock.Log.ConsoleError("The default usergroup could not be found. This may indicate a typo in the configuration file, or that the group was renamed or deleted.");
+				throw new Exception("The default usergroup could not be found.");
+            }
```
Can you indent using tabs here?

hakusaro commented on TShockAPI/DB/GroupManager.cs:
```diff
@@ -202,6 +202,21 @@ public GroupManager(IDbConnection db)
+			}
+
+			if(!GroupExists(TShock.Config.Settings.DefaultRegistrationGroupName))
+            {
```
Can you indent using tabs here?

hakusaro reviewed (changes requested):
I think this is a really good step to helping our support issues. I think, though, I'm a bit unsure if we should just have `Assert` and `Ensure` methods. `Ensure` seems to me like something that would create it if it didn't exist. I would prefer that `Ensure` did that, actually, else it should just be something like `Assert` just like the other method.

Fundamentally, though, I think if a player tries to join and their group is gone, you should also output an exception and tell someone how to resolve the issue, too.

Basically, I would say that this is good, but you'll probably want to convey more to the server operator. The `Ensure` method is more of a subset of `Assert` and they both try to accomplish similar things, so I'd suggest just combining them so that there's always an exception going to a server console. If the login fails because the group doesn't exist, that can be a separate thing.

Also, because this will probably cause someone some heart ache, I would suggest noting in the changelog exactly what will happen if groups are missing in which stages of the connection and startup process.

bartico6 commented:
My sincere apologies for the delay, I'll get around to it now.

bartico6 committed:
Fixes for PR 2397.

- GroupManager now validates core groups in the constructor.
- EnsureCoreGroupsPresent -> AssertCoreGroupsPresent.
- Fix indentiation in some places.
- Clarify the intent of this PR in CHANGELOG.md.

bartico6 committed:
Further clarification in CHANGELOG.

bartico6 committed:
Merge branch 'general-devel' into fix-invalid-groups

bartico6 commented:
This should be good now, I made the changes you requested & resolved the merge conflict that came up.

hakusaro commented on CHANGELOG.md:
```diff
@@ -42,6 +42,9 @@ This is the rolling changelog for TShock for Terraria. Use past tense when addin
   * `OnPlaceItemFrame`: The check was newly added.
   * `OnFoodPlatterTryPlacing`: The check was newly added.
+* Fixed errors on startup not being reported to console. (@bartico6)
+* The server now correctly disconnects players with missing groups instead of throwing an exception, stalling the connection (@bartico6)
```
```suggestion
  * The server now correctly disconnects players with missing groups instead of throwing an exception, stalling the connection (@bartico6)
```

hakusaro commented on CHANGELOG.md:
```diff
@@ -42,6 +42,9 @@ This is the rolling changelog for TShock for Terraria. Use past tense when addin
   * `OnFoodPlatterTryPlacing`: The check was newly added.
+* Fixed errors on startup not being reported to console. (@bartico6)
+* The server now correctly disconnects players with missing groups instead of throwing an exception, stalling the connection (@bartico6)
+* The server now rejects login attempts from players who would end up with a missing group. (@bartico6)
```
```suggestion
  * The server now rejects login attempts from players who would end up with a missing group. (@bartico6)
```

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -823,10 +823,16 @@ private static void AttemptLogin(CommandArgs args)
+					if (!TShock.Utils.AssertGroupValid(args.Player, group, false))
+					{
+						args.Player.SendErrorMessage("Login attempt failed - see the message above.");
+						return;
```
At this return path, wouldn't the user be left in an inconsistent state since they've been connected but everything else is broken?

bartico6 replied:
No, this is the login command. If they try to log in into an account that would break their groups this rejects the login attempt. But now I'm doubting my own code lol

hakusaro reviewed (changes requested):
Hmm, this still adds a method to TShock.Utils, but you marked my last comment as resolved. So I unresolved it.

I'm not quite sure what the status is if a user joins in a bad state. I guess it sucks but it's better than kicking them?

Why did the submodule change?

bartico6 commented:
I merged gendev because of a merge conflict so maybe that's why?
Also forgot to remove the method in utils after moving it to group manager it seems.

hakusaro commented:
> I merged gendev because of a merge conflict so maybe that's why?

![photo_2021-08-02 22 50 39](https://user-images.githubusercontent.com/532647/127964671-8b8e2566-519f-4607-ac3b-aeae0a3b51aa.jpeg)

hakusaro commented:
@bartico6 you still hacking on this or nah?

bartico6 commented:
Oh my, I forgot to update it again. I'll fix it in an hour, sincere apologies :c

bartico6 committed:
Remove the test method from Utils, re-route checks to GroupManager

bartico6 commented:
This time this *really* should be good. Please check :-)

hakusaro reviewed (changes requested):
This is almost ready for merge, but the submodule change will revert TSAPI back by 1 commit to the tree from DeathCradle's fork, not the tree post merge. Can you undo that change? Everything else lgtm!

bartico6 commented:
I'm not sure what I'm supposed to do here, I didn't make any changes to TSAPI myself, only did a merge from general-devel... should I merge gendev again hoping it evens out? should I forcibly checkout the previous submodule version?

bartico6 committed:
Checkout the same TSAPI as gen-dev.

bartico6 commented:
This explains what I did. As a sidenote it appears I.. checked out `disable-commands` in the work tree? I guess I was checking something and forgot to undo it, it has been a month. Sorry for the idiotic commits, we might want to squash it maybe?
![obraz](https://user-images.githubusercontent.com/3310937/128629022-b1356101-6c2f-4c57-8d93-9c758963c42e.png)

hakusaro reviewed (approved):
![lgtm_7](https://user-images.githubusercontent.com/532647/130729166-456eb481-c38c-481b-924d-f82662bd852b.gif)

QuiCM committed:
Merge branch 'general-devel' into fix-invalid-groups
